### PR TITLE
Feature/basis

### PR DESCRIPTION
**Pull request by Jammy2211**
A Basis is a list of light profiles used to model a Galaxy light profile, for example 10 Gaussian's:

```
gaussian_m = af.UniformPrior(lower_limit=0.0, upper_limit=1.0)
gaussian_c = af.UniformPrior(lower_limit=0.0, upper_limit=10.0)

gaussians = af.Collection([af.Model(ag.lp_linear.EllGaussian) for i in range(10)])

for i, gaussian in enumerate(gaussians):

    gaussian.centre = gaussians[0].centre
    gaussian.elliptical_comps = gaussians[0].elliptical_comps
    gaussian.sigma = (gaussian_m * i) + gaussian_c

bulge = af.Model(ag.lp_basis.Basis, light_profile_list=gaussians, regularization=ag.reg.Constant)

galaxy = af.Model(ag.Galaxy, redshift=0.5, bulge=bulge)

model = af.Collection(galaxies=af.Collection(galaxy=galaxy))
```

This PR implements the `Basis` class in the source code. 

A `Basis` can be regularized based on the light profiles it contains. This required the `Inversion` module to be extended to support lists of light profiles being linear solved for (with regularization), which makes up most the work in this PR.

**There is a lot of docs missing, which I will add in a future PR. There is another bigish refactor coming after this PR in the same parts of the code -- but this branch has had a lot of work done to it so I figure its best to get a review done now. **

### CODE REVIEW

File: autogalaxy/abstract_fit.py [hunk 1]
```diff
@@ -89,18 +85,25 @@ def linear_light_profile_intensity_dict(
...
         linear_light_profile_intensity_dict = {}
 
         for linear_obj_func in linear_obj_func_list:
 
-            linear_light_profile_intensity_dict[linear_obj_func.light_profile] = float(
-                self.inversion.reconstruction_dict[linear_obj_func]
-            )
+            reconstruction = self.inversion.reconstruction_dict[linear_obj_func]
+
+            for i, light_profile in enumerate(linear_obj_func.light_profile_list):
```
Comment (rhayes777) [inline autogalaxy/abstract_fit.py:102]: Use zip?

File: autogalaxy/analysis/model_util.py [hunk 2]
```diff
@@ -136,19 +136,21 @@ def set_upper_limit_of_pixelization_pixels_prior(
...
+                                log_str = (
+                                    "MODIFY BEFORE FIT -  A pixelization's pixel UniformPrior upper limit"
+                                    "was greater than the number of pixels in the mask. It has been "
+                                    "reduced to the number of pixels in the mask.\,"
+                                )
 
                                 if lower_limit > pixels_in_mask:
 
-                                    lower_limit = pixels_in_mask - 1
+                                    lower_limit = pixels_in_mask - 10
```
Comment (rhayes777) [inline autogalaxy/analysis/model_util.py:147]: Make this number a constant with a descriptive name

File: autogalaxy/plane/plane.py [hunk 3]
```diff
@@ -322,8 +323,27 @@ def contribution_map_list(self) -> List[aa.Array2D]:
...
+        Returns a bool specifying whether this fit object performs an inversion.
+
+        This is based on whether any of the galaxies in the `model_obj` have a `Pixelization` or `LightProfileLinear`
+        object, in which case an inversion is performed.
+
+        Returns
+        -------
+            A bool which is True if an inversion is performed.
+        """
+        if self.has(cls=aa.pix.Pixelization) or self.has(cls=LightProfileLinear):
```
Comment (rhayes777) [inline autogalaxy/plane/plane.py:337]: As the has method uses isinstance I think you can rewrite this:

```python
if self.has(cls=(aa.pix.Pixelization, LightProfileLinear)):

Comment (rhayes777) [reply autogalaxy/plane/plane.py:337]: Raised a PR to fix the type annotation on this

File: autogalaxy/plane/to_inversion.py [hunk 4]
```diff
@@ -1,62 +1,190 @@
...
-        if not self.plane.has(cls=LightProfileLinear):
+        self.plane = plane
+        self.dataset = dataset
+        self.data = data
+        self.noise_map = noise_map
+        self.w_tilde = w_tilde
+
+        if grid is not None:
+            self.grid = grid
+        elif dataset is not None:
```
Comment (rhayes777) [inline autogalaxy/plane/to_inversion.py:96]: Might be more readable as:

```python
try:
    self.grid = grid or dataset.grid
except AttributeError:
    self.grid = None

File: autogalaxy/profiles/light_profiles/basis.py [hunk 5]
```diff
@@ -0,0 +1,39 @@
...
+from typing import List, Optional
+
+import autoarray as aa
+
+from autogalaxy.profiles.light_profiles import light_profiles as lp
+from autogalaxy.profiles.light_profiles import light_profiles_linear as lp_linear
+
+
+class Basis(lp.LightProfile):
+    def __init__(
```
Comment (rhayes777) [inline autogalaxy/profiles/light_profiles/basis.py:11]: Add docs

File: test_autogalaxy/imaging/test_fit_imaging.py [hunk 6]
```diff
@@ -144,14 +136,30 @@ def test__noise_map__with_and_without_hyper_background(masked_imaging_7x7_no_blu
...
+    assert fit.figure_of_merit == pytest.approx(-75938.05, 1.0e-4)
+
+    basis = ag.lp_basis.Basis(
+        light_profile_list=[
+            ag.lp.EllSersic(intensity=1.0),
+            ag.lp.EllSersic(intensity=1.0),
+        ]
+    )
+
+    g0 = ag.Galaxy(redshift=0.5, bulge=basis)
```
Comment (rhayes777) [inline test_autogalaxy/imaging/test_fit_imaging.py:156]: This test is huge.

**Review by rhayes777 — APPROVED** [review]

File: autogalaxy/abstract_fit.py (hunk 1 above)
Comment (Jammy2211) [reply autogalaxy/abstract_fit.py:102]: Thats not how enumerate works...

**Pull request merged**